amethyst_utils/src/circular_buffer.rs

```diff
@@ -2,23 +2,40 @@
 
 use std::collections::VecDeque;
 
-///A CircularBuffer that drops the oldest element inserted when full.
+/// A CircularBuffer that drops the oldest element inserted when full.
+/// # Example
+///
+/// ```rust
+/// # use amethyst_utils::circular_buffer::CircularBuffer;
+/// # use std::collections::VecDeque;
+/// let mut buf = CircularBuffer::<u32>::new(2);
+/// assert_eq!(*buf.queue(), VecDeque::<u32>::from(vec![]));
+/// assert!(buf.push(1).is_none());
+/// assert_eq!(*buf.queue(), VecDeque::<u32>::from(vec![1]));
+/// assert!(buf.push(2).is_none());
+/// assert_eq!(*buf.queue(), VecDeque::<u32>::from(vec![1, 2]));
+/// assert!(buf.push(3).is_some());
+/// assert_eq!(*buf.queue(), VecDeque::<u32>::from(vec![2, 3]));
+/// assert_eq!(buf.capacity(), 2);
+/// ```
 pub struct CircularBuffer<A> {
     queue: VecDeque<A>,
+    cap: usize,
 }
 
 impl<A> CircularBuffer<A> {
     ///Creates a new CircularBuffer with fixed size
     pub fn new(size: usize) -> Self {
         CircularBuffer {
             queue: VecDeque::with_capacity(size),
+            cap: size,
         }
     }
 
     ///Add a value to the CircularBuffer
     ///Returns the popped value if the buffer is full
     pub fn push(&mut self, elem: A) -> Option<A> {
-        let out = if self.queue.len() == self.queue.capacity() {
+        let out = if self.queue.len() == self.cap {
```

Comment:
the queue capacity is dynamic right? (it grows when the buffer fills up)

Reply:
The thing is that `queue.capacity()` is unreliable. It is guaranteed to be AT LEAST what you specified, but can be greater than it. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dc6c7019a244ca8853df62a5728563de